cli: TestRemoveDeadReplicas failed

cli.TestRemoveDeadReplicas [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4149951&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4149951&tab=artifacts#/) on master @ [912964e02ddd951c77d4f71981ae18b3894e9084](https://github.com/cockroachdb/cockroach/commits/912964e02ddd951c77d4f71981ae18b3894e9084):


```
replica has lost quorum, recovering: r2:/System/NodeLiveness{-Max} [(n1,s1):1, (n3,s3):2, (n2,s2):3, (n4,s4):4, next=5, gen=6] -> r2:/System/NodeLiveness{-Max} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r3:/System/{NodeLivenessMax-tsd} [(n1,s1):1, (n3,s3):2, (n4,s4):3, (n2,s2):4, next=5, gen=6] -> r3:/System/{NodeLivenessMax-tsd} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r4:/System{/tsd-tse} [(n1,s1):1, (n3,s3):2, (n2,s2):3, (n4,s4):4, next=5, gen=6] -> r4:/System{/tsd-tse} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r5:/{Systemtse-Table/SystemConfigSpan/Start} [(n1,s1):1, (n4,s4):2, (n3,s3):3, (n2,s2):4, next=5, gen=6] -> r5:/{Systemtse-Table/SystemConfigSpan/Start} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r6:/Table/{SystemConfigSpan/Start-11} [(n3,s3):4, (n4,s4):2, (n2,s2):3, (n1,s1):5, next=6, gen=10] -> r6:/Table/{SystemConfigSpan/Start-11} [(n2,s2):6, next=7, gen=10]
replica has lost quorum, recovering: r7:/Table/1{1-2} [(n1,s1):1, (n4,s4):2, (n2,s2):3, (n3,s3):4, next=5, gen=6] -> r7:/Table/1{1-2} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r8:/Table/1{2-3} [(n3,s3):4, (n2,s2):2, (n4,s4):3, (n1,s1):5, next=6, gen=10] -> r8:/Table/1{2-3} [(n2,s2):6, next=7, gen=10]
replica has lost quorum, recovering: r9:/Table/1{3-4} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=6] -> r9:/Table/1{3-4} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r10:/Table/1{4-5} [(n1,s1):1, (n3,s3):2, (n2,s2):3, (n4,s4):4, next=5, gen=6] -> r10:/Table/1{4-5} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r11:/Table/1{5-6} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=6] -> r11:/Table/1{5-6} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r12:/Table/1{6-7} [(n1,s1):1, (n4,s4):2, (n3,s3):3, (n2,s2):4, next=5, gen=6] -> r12:/Table/1{6-7} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r13:/Table/1{7-8} [(n1,s1):1, (n4,s4):2, (n3,s3):3, (n2,s2):4, next=5, gen=6] -> r13:/Table/1{7-8} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r14:/Table/1{8-9} [(n1,s1):1, (n3,s3):2, (n2,s2):3, (n4,s4):4, next=5, gen=6] -> r14:/Table/1{8-9} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r15:/Table/{19-20} [(n1,s1):1, (n4,s4):2, (n2,s2):3, (n3,s3):4, next=5, gen=6] -> r15:/Table/{19-20} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r16:/Table/2{0-1} [(n1,s1):1, (n3,s3):2, (n4,s4):3, (n2,s2):4, next=5, gen=6] -> r16:/Table/2{0-1} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r17:/Table/2{1-2} [(n1,s1):1, (n3,s3):2, (n2,s2):3, (n4,s4):4, next=5, gen=6] -> r17:/Table/2{1-2} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r18:/Table/2{2-3} [(n1,s1):1, (n2,s2):2, (n4,s4):3, (n3,s3):4, next=5, gen=6] -> r18:/Table/2{2-3} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r19:/Table/2{3-4} [(n1,s1):1, (n4,s4):2, (n3,s3):3, (n2,s2):4, next=5, gen=6] -> r19:/Table/2{3-4} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r20:/Table/2{4-5} [(n1,s1):1, (n3,s3):2, (n4,s4):3, (n2,s2):4, next=5, gen=6] -> r20:/Table/2{4-5} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r21:/Table/2{5-6} [(n1,s1):1, (n4,s4):2, (n2,s2):3, (n3,s3):4, next=5, gen=6] -> r21:/Table/2{5-6} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r22:/Table/2{6-7} [(n1,s1):1, (n4,s4):2, (n2,s2):3, (n3,s3):4, next=5, gen=6] -> r22:/Table/2{6-7} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r23:/Table/2{7-8} [(n1,s1):1, (n4,s4):2, (n3,s3):3, (n2,s2):4, next=5, gen=6] -> r23:/Table/2{7-8} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r24:/Table/2{8-9} [(n3,s3):4, (n2,s2):2, (n4,s4):3, (n1,s1):5, next=6, gen=10] -> r24:/Table/2{8-9} [(n2,s2):6, next=7, gen=10]
replica has lost quorum, recovering: r25:/{Table/29-NamespaceTable/30} [(n1,s1):1, (n4,s4):2, (n2,s2):3, (n3,s3):4, next=5, gen=6] -> r25:/{Table/29-NamespaceTable/30} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r26:/NamespaceTable/{30-Max} [(n1,s1):1, (n4,s4):2, (n2,s2):3, (n3,s3):4, next=5, gen=6] -> r26:/NamespaceTable/{30-Max} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r27:/{NamespaceTable/Max-Table/32} [(n1,s1):1, (n2,s2):2, (n4,s4):3, (n3,s3):4, next=5, gen=6] -> r27:/{NamespaceTable/Max-Table/32} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r28:/Table/3{2-3} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=6] -> r28:/Table/3{2-3} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r29:/Table/3{3-4} [(n1,s1):1, (n3,s3):2, (n2,s2):3, (n4,s4):4, next=5, gen=6] -> r29:/Table/3{3-4} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r30:/Table/3{4-5} [(n1,s1):1, (n2,s2):2, (n4,s4):3, (n3,s3):4, next=5, gen=6] -> r30:/Table/3{4-5} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r31:/Table/3{5-6} [(n1,s1):1, (n3,s3):2, (n2,s2):3, (n4,s4):4, next=5, gen=6] -> r31:/Table/3{5-6} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r32:/Table/3{6-7} [(n1,s1):1, (n4,s4):2, (n2,s2):3, (n3,s3):4, next=5, gen=6] -> r32:/Table/3{6-7} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r33:/Table/3{7-8} [(n1,s1):1, (n4,s4):2, (n2,s2):3, (n3,s3):4, next=5, gen=6] -> r33:/Table/3{7-8} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r34:/Table/3{8-9} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=6] -> r34:/Table/3{8-9} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r35:/Table/{39-40} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=6] -> r35:/Table/{39-40} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r36:/Table/4{0-1} [(n1,s1):1, (n4,s4):2, (n2,s2):3, (n3,s3):4, next=5, gen=6] -> r36:/Table/4{0-1} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r37:/Table/4{1-2} [(n1,s1):1, (n4,s4):2, (n3,s3):3, (n2,s2):4, next=5, gen=6] -> r37:/Table/4{1-2} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r38:/Table/4{2-3} [(n1,s1):1, (n3,s3):2, (n4,s4):3, (n2,s2):4, next=5, gen=6] -> r38:/Table/4{2-3} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r39:/Table/4{3-4} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=6] -> r39:/Table/4{3-4} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r40:/Table/4{4-5} [(n2,s2):4, (n4,s4):2, (n3,s3):3, (n1,s1):5, next=6, gen=10] -> r40:/Table/4{4-5} [(n2,s2):6, next=7, gen=10]
replica has lost quorum, recovering: r41:/Table/4{5-6} [(n1,s1):1, (n4,s4):2, (n3,s3):3, (n2,s2):4, next=5, gen=6] -> r41:/Table/4{5-6} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r42:/Table/4{6-7} [(n1,s1):1, (n3,s3):2, (n4,s4):3, (n2,s2):4, next=5, gen=6] -> r42:/Table/4{6-7} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r43:/{Table/47-Max} [(n1,s1):1, (n3,s3):2, (n4,s4):3, (n2,s2):4, next=5, gen=6] -> r43:/{Table/47-Max} [(n2,s2):5, next=6, gen=6]
aborting intent: /Local/Range/System/tsd/RangeDescriptor (txn 4435d938-2f83-426c-a270-6aff4c966650)
Scanning replicas on store cluster_id:00578a7c-aef7-4181-b427-3f8d8aaa401a node_id:2 store_id:2  for dead peers []

id	is_live	replicas	is_decommissioning	membership	is_draining
3	false	43	true	decommissioning	false
4	false	43	true	decommissioning	false
    debug_test.go:391: expected replicas on {1,2,5,6} but got {1,2,5}
    --- FAIL: TestRemoveDeadReplicas/2/4/r=4 (26.39s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/server @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRemoveDeadReplicas.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-12480

cli.TestRemoveDeadReplicas [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4186518&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4186518&tab=artifacts#/) on master @ [e1068d77afbd39b162978281c9da7cbea49c1c3a](https://github.com/cockroachdb/cockroach/commits/e1068d77afbd39b162978281c9da7cbea49c1c3a):


```
replica has lost quorum, recovering: r2:/System/NodeLiveness{-Max} [(n1,s1):1, (n2,s2):2, (n4,s4):3, (n3,s3):4, next=5, gen=6] -> r2:/System/NodeLiveness{-Max} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r3:/System/{NodeLivenessMax-tsd} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=6] -> r3:/System/{NodeLivenessMax-tsd} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r4:/System{/tsd-tse} [(n1,s1):1, (n4,s4):2, (n3,s3):3, (n2,s2):4, next=5, gen=6] -> r4:/System{/tsd-tse} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r5:/{Systemtse-Table/SystemConfigSpan/Start} [(n1,s1):1, (n3,s3):2, (n4,s4):3, (n2,s2):4, next=5, gen=6] -> r5:/{Systemtse-Table/SystemConfigSpan/Start} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r6:/Table/{SystemConfigSpan/Start-11} [(n3,s3):4, (n4,s4):2, (n2,s2):3, (n1,s1):5, next=6, gen=10] -> r6:/Table/{SystemConfigSpan/Start-11} [(n2,s2):6, next=7, gen=10]
replica has lost quorum, recovering: r7:/Table/1{1-2} [(n1,s1):1, (n4,s4):2, (n2,s2):3, (n3,s3):4, next=5, gen=6] -> r7:/Table/1{1-2} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r8:/Table/1{2-3} [(n1,s1):1, (n3,s3):2, (n2,s2):3, (n4,s4):4, next=5, gen=6] -> r8:/Table/1{2-3} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r9:/Table/1{3-4} [(n1,s1):1, (n3,s3):2, (n2,s2):3, (n4,s4):4, next=5, gen=6] -> r9:/Table/1{3-4} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r10:/Table/1{4-5} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=6] -> r10:/Table/1{4-5} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r11:/Table/1{5-6} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=6] -> r11:/Table/1{5-6} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r12:/Table/1{6-7} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=6] -> r12:/Table/1{6-7} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r13:/Table/1{7-8} [(n2,s2):4, (n3,s3):2, (n4,s4):3, (n1,s1):5, next=6, gen=10] -> r13:/Table/1{7-8} [(n2,s2):6, next=7, gen=10]
replica has lost quorum, recovering: r14:/Table/1{8-9} [(n1,s1):1, (n2,s2):2, (n4,s4):3, (n3,s3):4, next=5, gen=6] -> r14:/Table/1{8-9} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r15:/Table/{19-20} [(n1,s1):1, (n3,s3):2, (n2,s2):3, (n4,s4):4, next=5, gen=6] -> r15:/Table/{19-20} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r16:/Table/2{0-1} [(n1,s1):1, (n4,s4):2, (n3,s3):3, (n2,s2):4, next=5, gen=6] -> r16:/Table/2{0-1} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r17:/Table/2{1-2} [(n1,s1):1, (n4,s4):2, (n2,s2):3, (n3,s3):4, next=5, gen=6] -> r17:/Table/2{1-2} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r18:/Table/2{2-3} [(n1,s1):1, (n4,s4):2, (n2,s2):3, (n3,s3):4, next=5, gen=6] -> r18:/Table/2{2-3} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r19:/Table/2{3-4} [(n1,s1):1, (n4,s4):2, (n3,s3):3, (n2,s2):4, next=5, gen=6] -> r19:/Table/2{3-4} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r20:/Table/2{4-5} [(n1,s1):1, (n4,s4):2, (n3,s3):3, (n2,s2):4, next=5, gen=6] -> r20:/Table/2{4-5} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r21:/Table/2{5-6} [(n1,s1):1, (n2,s2):2, (n4,s4):3, (n3,s3):4, next=5, gen=6] -> r21:/Table/2{5-6} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r22:/Table/2{6-7} [(n4,s4):4, (n3,s3):2, (n2,s2):3, (n1,s1):5, next=6, gen=10] -> r22:/Table/2{6-7} [(n2,s2):6, next=7, gen=10]
replica has lost quorum, recovering: r23:/Table/2{7-8} [(n1,s1):1, (n3,s3):2, (n4,s4):3, (n2,s2):4, next=5, gen=6] -> r23:/Table/2{7-8} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r24:/Table/2{8-9} [(n1,s1):1, (n2,s2):2, (n4,s4):3, (n3,s3):4, next=5, gen=6] -> r24:/Table/2{8-9} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r25:/{Table/29-NamespaceTable/30} [(n1,s1):1, (n4,s4):2, (n3,s3):3, (n2,s2):4, next=5, gen=6] -> r25:/{Table/29-NamespaceTable/30} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r26:/NamespaceTable/{30-Max} [(n1,s1):1, (n3,s3):2, (n2,s2):3, (n4,s4):4, next=5, gen=6] -> r26:/NamespaceTable/{30-Max} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r27:/{NamespaceTable/Max-Table/32} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=6] -> r27:/{NamespaceTable/Max-Table/32} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r28:/Table/3{2-3} [(n1,s1):1, (n3,s3):2, (n4,s4):3, (n2,s2):4, next=5, gen=6] -> r28:/Table/3{2-3} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r29:/Table/3{3-4} [(n1,s1):1, (n4,s4):2, (n3,s3):3, (n2,s2):4, next=5, gen=6] -> r29:/Table/3{3-4} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r30:/Table/3{4-5} [(n1,s1):1, (n4,s4):2, (n2,s2):3, (n3,s3):4, next=5, gen=6] -> r30:/Table/3{4-5} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r31:/Table/3{5-6} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=6] -> r31:/Table/3{5-6} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r32:/Table/3{6-7} [(n1,s1):1, (n3,s3):2, (n2,s2):3, (n4,s4):4, next=5, gen=6] -> r32:/Table/3{6-7} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r33:/Table/3{7-8} [(n1,s1):1, (n4,s4):2, (n3,s3):3, (n2,s2):4, next=5, gen=6] -> r33:/Table/3{7-8} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r34:/Table/3{8-9} [(n1,s1):1, (n4,s4):2, (n2,s2):3, (n3,s3):4, next=5, gen=6] -> r34:/Table/3{8-9} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r35:/Table/{39-40} [(n1,s1):1, (n4,s4):2, (n3,s3):3, (n2,s2):4, next=5, gen=6] -> r35:/Table/{39-40} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r36:/Table/4{0-1} [(n1,s1):1, (n4,s4):2, (n3,s3):3, (n2,s2):4, next=5, gen=6] -> r36:/Table/4{0-1} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r37:/Table/4{1-2} [(n1,s1):1, (n2,s2):2, (n4,s4):3, (n3,s3):4, next=5, gen=6] -> r37:/Table/4{1-2} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r38:/Table/4{2-3} [(n1,s1):1, (n4,s4):2, (n2,s2):3, (n3,s3):4, next=5, gen=6] -> r38:/Table/4{2-3} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r39:/Table/4{3-4} [(n1,s1):1, (n3,s3):2, (n4,s4):3, (n2,s2):4, next=5, gen=6] -> r39:/Table/4{3-4} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r40:/Table/4{4-5} [(n1,s1):1, (n3,s3):2, (n2,s2):3, (n4,s4):4, next=5, gen=6] -> r40:/Table/4{4-5} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r41:/Table/4{5-6} [(n1,s1):1, (n2,s2):2, (n4,s4):3, (n3,s3):4, next=5, gen=6] -> r41:/Table/4{5-6} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r42:/Table/4{6-7} [(n1,s1):1, (n2,s2):2, (n4,s4):3, (n3,s3):4, next=5, gen=6] -> r42:/Table/4{6-7} [(n2,s2):5, next=6, gen=6]
replica has lost quorum, recovering: r43:/{Table/47-Max} [(n1,s1):1, (n4,s4):2, (n2,s2):3, (n3,s3):4, next=5, gen=6] -> r43:/{Table/47-Max} [(n2,s2):5, next=6, gen=6]
aborting intent: /Local/Range/System/tsd/RangeDescriptor (txn 1f622e19-4a93-4ede-9ed3-202a9adde8ef)
Scanning replicas on store cluster_id:1f678502-228d-44a8-9d7a-f791bb7fb4ed node_id:2 store_id:2  for dead peers []

id	is_live	replicas	is_decommissioning	membership	is_draining
3	false	43	true	decommissioning	false
4	false	43	true	decommissioning	false
    debug_test.go:391: expected replicas on {1,2,5,6} but got {1,2,6}
    --- FAIL: TestRemoveDeadReplicas/2/4/r=4 (24.82s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRemoveDeadReplicas.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

cli.TestRemoveDeadReplicas [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4614851&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4614851&tab=artifacts#/) on master @ [d84f07e856dbb3d62f0336f4e8439280e2a018ed](https://github.com/cockroachdb/cockroach/commits/d84f07e856dbb3d62f0336f4e8439280e2a018ed):


```
=== RUN   TestRemoveDeadReplicas
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/511bb7a3a798948c9c27fde89a1fae1e/logTestRemoveDeadReplicas1858003926
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestRemoveDeadReplicas
    debug_test.go:414: -- test log scope end --
--- FAIL: TestRemoveDeadReplicas (24.79s)
=== RUN   TestRemoveDeadReplicas/2/4/r=4
    debug_test.go:234: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_WRITE_TOO_OLD - WriteTooOld flag converted to WriteTooOldError): "unnamed" meta={id=7fd6ce0d key=/Local/Range/System/tsd/RangeDescriptor pri=0.01740530 epo=0 ts=1647670503.023559277,1 min=1647670503.002154173,0 seq=1} lock=true stat=PENDING rts=1647670503.002154173,0 wto=false gul=1647670503.502154173,0
    --- FAIL: TestRemoveDeadReplicas/2/4/r=4 (6.72s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRemoveDeadReplicas.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/cli/debug.go
+++ b/pkg/cli/debug.go
@@ -1371,6 +1371,15 @@ func removeDeadReplicas(
 			batch.Close()
 			return nil, err
 		}
+		// Write the new replica ID to RaftReplicaIDKey.
+		replicas := desc.Replicas().Descriptors()
+		if len(replicas) != 1 {
+			return nil, errors.Errorf("expected 1 replica, got %v", replicas)
+		}
+		if err := sl.SetRaftReplicaID(ctx, batch, replicas[0].ReplicaID); err != nil {
+			return nil, errors.Wrapf(err, "failed to write new replica ID for range %d", desc.RangeID)
+		}
+		// Update MVCC stats.
 		if err := sl.SetMVCCStats(ctx, batch, &ms); err != nil {
 			return nil, errors.Wrap(err, "updating MVCCStats")
 		}

--- a/pkg/kv/kvserver/loqrecovery/apply.go
+++ b/pkg/kv/kvserver/loqrecovery/apply.go
@@ -293,6 +293,11 @@ func applyReplicaUpdate(
 	report.OldReplica, _ = report.RemovedReplicas.RemoveReplica(
 		update.NewReplica.NodeID, update.NewReplica.StoreID)
 
+	// Persist the new replica ID.
+	if err := sl.SetRaftReplicaID(ctx, readWriter, update.NewReplica.ReplicaID); err != nil {
+		return PrepareReplicaReport{}, errors.Wrap(err, "setting new replica ID")
+	}
+
 	// Refresh stats
 	if err := sl.SetMVCCStats(ctx, readWriter, &ms); err != nil {
 		return PrepareReplicaReport{}, errors.Wrap(err, "updating MVCCStats")

--- a/pkg/kv/kvserver/loqrecovery/recovery_env_test.go
+++ b/pkg/kv/kvserver/loqrecovery/recovery_env_test.go
@@ -81,6 +81,7 @@ type storeView struct {
 	StoreID roachpb.StoreID `yaml:"StoreID"`
 
 	Descriptors []storeDescriptorView `yaml:"Descriptors"`
+	LocalData   []localDataView       `yaml:"LocalData"`
 }
 
 // storeDescriptorView contains important fields from the range
@@ -117,6 +118,12 @@ func (r replicaDescriptorView) asReplicaDescriptor() roachpb.ReplicaDescriptor {
 		ReplicaID: r.ReplicaID,
 		Type:      r.ReplicaType,
 	}
+}
+
+// localDataView contains interesting local store data for each range.
+type localDataView struct {
+	RangeID       roachpb.RangeID `yaml:"RangeID"`
+	RaftReplicaID int             `yaml:"RaftReplicaID"`
 }
 
 // Store with its owning NodeID for easier grouping by owning nodes.
@@ -195,7 +202,8 @@ func (e *quorumRecoveryEnv) handleReplicationData(t *testing.T, d datadriven.Tes
 			replica.NodeID = roachpb.NodeID(replica.StoreID)
 		}
 
-		key, desc, replicaState, hardState, raftLog := buildReplicaDescriptorFromTestData(t, replica)
+		replicaID, key, desc, replicaState, hardState, raftLog :=
+			buildReplicaDescriptorFromTestData(t, replica)
 
 		eng := e.getOrCreateStore(ctx, t, replica.StoreID, replica.NodeID)
 		if err = storage.MVCCPutProto(ctx, eng, nil, key, clock.Now(), nil, /* txn */
@@ -209,6 +217,9 @@ func (e *quorumRecoveryEnv) handleReplicationData(t *testing.T, d datadriven.Tes
 		}
 		if err := sl.SetHardState(ctx, eng, hardState); err != nil {
 			t.Fatalf("failed to save raft hard state: %v", err)
+		}
+		if err := sl.SetRaftReplicaID(ctx, eng, replicaID); err != nil {
+			t.Fatalf("failed to set raft replica ID: %v", err)
 		}
 		for i, entry := range raftLog {
 			value, err := protoutil.Marshal(&entry)
@@ -227,6 +238,7 @@ func (e *quorumRecoveryEnv) handleReplicationData(t *testing.T, d datadriven.Tes
 func buildReplicaDescriptorFromTestData(
 	t *testing.T, replica testReplicaInfo,
 ) (
+	roachpb.ReplicaID,
 	roachpb.Key,
 	roachpb.RangeDescriptor,
 	kvserverpb.ReplicaState,
@@ -239,12 +251,16 @@ func buildReplicaDescriptorFromTestData(
 	endKey := parsePrettyKey(t, replica.EndKey)
 	key := keys.RangeDescriptorKey(startKey)
 	var replicas []roachpb.ReplicaDescriptor
+	var replicaID roachpb.ReplicaID
 	maxReplicaID := replica.Replicas[0].ReplicaID
 	for _, r := range replica.Replicas {
 		if r.ReplicaID > maxReplicaID {
 			maxReplicaID = r.ReplicaID
 		}
 		replicas = append(replicas, r.asReplicaDescriptor())
+		if r.NodeID == replica.NodeID && r.StoreID == replica.StoreID {
+			replicaID = r.ReplicaID
+		}
 	}
 	if replica.Generation == 0 {
 		replica.Generation = roachpb.RangeGeneration(maxReplicaID)
@@ -292,7 +308,7 @@ func buildReplicaDescriptorFromTestData(
 		entry := raftLogFromPendingDescriptorUpdate(t, replica, u, desc, uint64(i))
 		raftLog = append(raftLog, enginepb.MVCCMetadata{RawBytes: entry.RawBytes})
 	}
-	return key, desc, replicaState, hardState, raftLog
+	return replicaID, key, desc, replicaState, hardState, raftLog
 }
 
 func raftLogFromPendingDescriptorUpdate(
@@ -520,10 +536,21 @@ func (e *quorumRecoveryEnv) handleDumpStore(t *testing.T, d datadriven.TestData)
 	var storesView []storeView
 	for _, storeID := range stores {
 		var descriptorViews []storeDescriptorView
+		var localDataViews []localDataView
 		store := e.stores[storeID]
 		err := kvserver.IterateRangeDescriptorsFromDisk(ctx, store.engine,
 			func(desc roachpb.RangeDescriptor) error {
 				descriptorViews = append(descriptorViews, descriptorView(desc))
+
+				sl := stateloader.Make(desc.RangeID)
+				raftReplicaID, _, err := sl.LoadRaftReplicaID(ctx, store.engine)
+				if err != nil {
+					t.Fatalf("failed to load Raft replica ID: %v", err)
+				}
+				localDataViews = append(localDataViews, localDataView{
+					RangeID:       desc.RangeID,
+					RaftReplicaID: int(raftReplicaID.ReplicaID),
+				})
 				return nil
 			})
 		if err != nil {
@@ -533,6 +560,7 @@ func (e *quorumRecoveryEnv) handleDumpStore(t *testing.T, d datadriven.TestData)
 			NodeID:      e.stores[storeID].nodeID,
 			StoreID:     storeID,
 			Descriptors: descriptorViews,
+			LocalData:   localDataViews,
 		})
 	}
 	out, err := yaml.Marshal(storesView)

--- a/pkg/kv/kvserver/loqrecovery/testdata/learners_lose
+++ b/pkg/kv/kvserver/loqrecovery/testdata/learners_lose
@@ -61,6 +61,9 @@ dump-store stores=(1,2)
     StartKey: /Min
     Replicas:
     - Replica: {NodeID: 1, StoreID: 1, ReplicaID: 16}
+  LocalData:
+  - RangeID: 1
+    RaftReplicaID: 16
 - NodeID: 2
   StoreID: 2
   Descriptors:
@@ -72,6 +75,9 @@ dump-store stores=(1,2)
     - Replica: {NodeID: 3, StoreID: 3, ReplicaID: 3}
     - Replica: {NodeID: 4, StoreID: 4, ReplicaID: 4}
     - Replica: {NodeID: 5, StoreID: 5, ReplicaID: 5}
+  LocalData:
+  - RangeID: 1
+    RaftReplicaID: 2
 
 # Second use case where we can't make a decision and fail keyspace coverage as
 # only a single learner is left, there is no way to recover.

--- a/pkg/kv/kvserver/loqrecovery/testdata/max_applied_voter_wins
+++ b/pkg/kv/kvserver/loqrecovery/testdata/max_applied_voter_wins
@@ -59,6 +59,9 @@ dump-store stores=(1,2)
     StartKey: /Min
     Replicas:
     - Replica: {NodeID: 1, StoreID: 1, ReplicaID: 16}
+  LocalData:
+  - RangeID: 1
+    RaftReplicaID: 16
 - NodeID: 2
   StoreID: 2
   Descriptors:
@@ -70,6 +73,9 @@ dump-store stores=(1,2)
     - Replica: {NodeID: 3, StoreID: 3, ReplicaID: 3}
     - Replica: {NodeID: 4, StoreID: 4, ReplicaID: 4}
     - Replica: {NodeID: 5, StoreID: 5, ReplicaID: 5}
+  LocalData:
+  - RangeID: 1
+    RaftReplicaID: 2
 
 dump-events stores=(1,2)
 ----
@@ -184,6 +190,11 @@ dump-store stores=(1,2,5,6)
     - Replica: {NodeID: 1, StoreID: 1, ReplicaID: 1}
     - Replica: {NodeID: 5, StoreID: 5, ReplicaID: 6}
     - Replica: {NodeID: 6, StoreID: 6, ReplicaID: 7}
+  LocalData:
+  - RangeID: 1
+    RaftReplicaID: 1
+  - RangeID: 2
+    RaftReplicaID: 1
 - NodeID: 2
   StoreID: 2
   Descriptors:
@@ -193,6 +204,9 @@ dump-store stores=(1,2,5,6)
     - Replica: {NodeID: 2, StoreID: 2, ReplicaID: 2}
     - Replica: {NodeID: 3, StoreID: 3, ReplicaID: 3}
     - Replica: {NodeID: 4, StoreID: 4, ReplicaID: 4}
+  LocalData:
+  - RangeID: 1
+    RaftReplicaID: 2
 - NodeID: 5
   StoreID: 5
   Descriptors:
@@ -208,6 +222,11 @@ dump-store stores=(1,2,5,6)
     - Replica: {NodeID: 1, StoreID: 1, ReplicaID: 1}
     - Replica: {NodeID: 5, StoreID: 5, ReplicaID: 6}
     - Replica: {NodeID: 6, StoreID: 6, ReplicaID: 7}
+  LocalData:
+  - RangeID: 1
+    RaftReplicaID: 6
+  - RangeID: 2
+    RaftReplicaID: 6
 - NodeID: 6
   StoreID: 6
   Descriptors:
@@ -223,3 +242,8 @@ dump-store stores=(1,2,5,6)
     - Replica: {NodeID: 1, StoreID: 1, ReplicaID: 1}
     - Replica: {NodeID: 5, StoreID: 5, ReplicaID: 6}
     - Replica: {NodeID: 6, StoreID: 6, ReplicaID: 7}
+  LocalData:
+  - RangeID: 1
+    RaftReplicaID: 7
+  - RangeID: 2
+    RaftReplicaID: 7

--- a/pkg/kv/kvserver/loqrecovery/testdata/max_store_voter_wins
+++ b/pkg/kv/kvserver/loqrecovery/testdata/max_store_voter_wins
@@ -64,6 +64,9 @@ dump-store stores=(1,2)
     - Replica: {NodeID: 3, StoreID: 3, ReplicaID: 3}
     - Replica: {NodeID: 4, StoreID: 4, ReplicaID: 4}
     - Replica: {NodeID: 5, StoreID: 5, ReplicaID: 5}
+  LocalData:
+  - RangeID: 1
+    RaftReplicaID: 1
 - NodeID: 2
   StoreID: 2
   Descriptors:
@@ -71,3 +74,6 @@ dump-store stores=(1,2)
     StartKey: /Min
     Replicas:
     - Replica: {NodeID: 2, StoreID: 2, ReplicaID: 16}
+  LocalData:
+  - RangeID: 1
+    RaftReplicaID: 16

--- a/pkg/kv/kvserver/loqrecovery/testdata/no_change_when_no_dead_peers
+++ b/pkg/kv/kvserver/loqrecovery/testdata/no_change_when_no_dead_peers
@@ -58,6 +58,9 @@ dump-store stores=(1,2,3)
     - Replica: {NodeID: 1, StoreID: 1, ReplicaID: 1}
     - Replica: {NodeID: 2, StoreID: 2, ReplicaID: 2}
     - Replica: {NodeID: 3, StoreID: 3, ReplicaID: 3}
+  LocalData:
+  - RangeID: 1
+    RaftReplicaID: 1
 - NodeID: 2
   StoreID: 2
   Descriptors:
@@ -67,6 +70,9 @@ dump-store stores=(1,2,3)
     - Replica: {NodeID: 1, StoreID: 1, ReplicaID: 1}
     - Replica: {NodeID: 2, StoreID: 2, ReplicaID: 2}
     - Replica: {NodeID: 3, StoreID: 3, ReplicaID: 3}
+  LocalData:
+  - RangeID: 1
+    RaftReplicaID: 2
 - NodeID: 3
   StoreID: 3
   Descriptors:
@@ -76,3 +82,6 @@ dump-store stores=(1,2,3)
     - Replica: {NodeID: 1, StoreID: 1, ReplicaID: 1}
     - Replica: {NodeID: 2, StoreID: 2, ReplicaID: 2}
     - Replica: {NodeID: 3, StoreID: 3, ReplicaID: 3}
+  LocalData:
+  - RangeID: 1
+    RaftReplicaID: 3

--- a/pkg/kv/kvserver/loqrecovery/testdata/no_change_when_quorum
+++ b/pkg/kv/kvserver/loqrecovery/testdata/no_change_when_quorum
@@ -47,6 +47,9 @@ dump-store stores=(1,2)
     - Replica: {NodeID: 1, StoreID: 1, ReplicaID: 1}
     - Replica: {NodeID: 2, StoreID: 2, ReplicaID: 2}
     - Replica: {NodeID: 3, StoreID: 3, ReplicaID: 3}
+  LocalData:
+  - RangeID: 1
+    RaftReplicaID: 1
 - NodeID: 2
   StoreID: 2
   Descriptors:
@@ -56,3 +59,6 @@ dump-store stores=(1,2)
     - Replica: {NodeID: 1, StoreID: 1, ReplicaID: 1}
     - Replica: {NodeID: 2, StoreID: 2, ReplicaID: 2}
     - Replica: {NodeID: 3, StoreID: 3, ReplicaID: 3}
+  LocalData:
+  - RangeID: 1
+    RaftReplicaID: 2

